lightning-data-persister/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+extern crate lightning;
+extern crate bitcoin;
+extern crate libc;
+
+use bitcoin::hash_types::{BlockHash, Txid};
+use bitcoin::hashes::hex::{ToHex, FromHex};
+use lightning::chain::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr};
+use lightning::chain::keysinterface::ChannelKeys;
+use lightning::chain::transaction::OutPoint;
+use lightning::ln::data_persister::ChannelDataPersister;
+use lightning::util::ser::{Writeable, Readable};
+use std::collections::HashMap;
+use std::fs;
+use std::io::{Error, ErrorKind, Cursor};
+use std::marker::PhantomData;
+use std::path::Path;
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::AsRawFd;
+
+/// FilesystemPersister can persist channel data on disk on Linux machines, where
+/// each channel's data is stored in a file named after its funding outpoint.
+///
+/// Warning: this module does the best it can with calls to persist data, but it
+/// can only guarantee that the data is passed to the drive. It is up to the
+/// drive manufacturers to do the actual persistence properly, which they often
+/// don't (especially on consumer-grade hardware). Therefore, it is up to the
+/// user to validate their entire storage stack, to ensure the writes are
+/// persistent.
+/// Corollary: especially when dealing with larger amounts of money, it is best
+/// practice to have multiple channel data backups and not rely only on the
+/// FilesystemPersister.
+pub struct FilesystemPersister<ChanSigner: ChannelKeys + Readable + Writeable> {
+	path_to_channel_data: String,
+	phantom: PhantomData<ChanSigner>, // TODO: is there a way around this?
+}
+
+impl<ChanSigner: ChannelKeys + Readable + Writeable> FilesystemPersister<ChanSigner> {
+	/// Initialize a new FilesystemPersister and set the path to the individual channels'
+	/// files.
+	pub fn new(path_to_channel_data: String) -> Self {
+		return Self {
+			path_to_channel_data,
+			phantom: PhantomData,
+		}
+	}
+
+	fn get_full_filepath(&self, funding_txo: OutPoint) -> String {
+		let dir_path = Path::new(&self.path_to_channel_data);
+		let mut path_buf = dir_path.to_path_buf();
+		path_buf.push(format!("{}_{}", funding_txo.txid.to_hex(), funding_txo.index));
+		path_buf.to_str().unwrap().to_string()
+	}
+
+	// Utility to write a file to disk.
+	// Note: we may eventually want to change this model to a queue-based system, such that
+	// writes are queued to a separate writer thread. This would improve performance
+	// since syncing to disk can sometimes take 100s of milliseconds.
+	fn write_channel_data(&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChanSigner>) -> std::io::Result<()> {
+		// Do a crazy dance with lots of fsync()s to be overly cautious here...
```

Comment:
It would make things more testable if the file system manipulation was moved to a generic utility function that took a byte array.  That way you could unit-test the file system dance independently.  

Right now the only test in this crate reads more like an integration test.  That's fine, but it's good to also have smaller tests to be able to focus on details such as error conditions.

The `ChannelMonitor` is only a few KB, so should be OK to serialize in memory first.

Reply:
What extra cases do you think should be covered?

Reply:
The main case to test is when the temporary file exists.  This would simulate a failure just before the rename.

Might also be good to test what happens if the file or directory are not writable.  The main concern is if the code somehow swallows errors.

Reply:
@TheBlueMatt wondering, would it be a potential performance issue to implement this idea (serializing the `ChannelMonitor` in-memory first)? I only question because the monitors can grow without bound, and could potentially have tons of updates per second.

I'd agree these cases should be tested, though. Will add tests.

Reply:
Hmm, right, `ChannelMonitors` grow with old state, so they're likely not always only a few KB - practically, pre-serializing will almost certainly be a performance *improvement* (fewer syscalls to write), but at the cost of doubling client memory, and probably fragmenting the heap a good bit, so not a great idea for large nodes or non-desktop/server platforms. We could totally still make it generic across serializable objects, though.